File: src/trace.zig
//! Instruction trace: one JSON object per executed instruction, on stdout.
//!
//! Format (fields in a fixed order so a diff between two runs is readable):
//!
//!   {"step":3,"pc":"0x40080006","raw":"0x0020c0","len":3,"op":"memw","a":["0x0", ...]}
//!
//! It is JSON *Lines* — one self-contained object per line, no enclosing array — so it can
//! be streamed, grepped, and piped into `jq` while the run is still going.
//!
//! The formatting is done by hand into a fixed-size buffer rather than through
//! `std.json`. Two reasons: the output shape is fixed and trivial, and doing it this way
//! keeps the tracer allocation-free and the formatting logic unit-testable as a pure
//! function (`formatLine`) with no I/O.

const std = @import("std");
const decode = @import("cpu/decode.zig");

/// Worst case: ~90 bytes of scaffolding plus 16 registers at up to 13 bytes each.
pub const line_buffer_size = 512;

pub const Tracer = struct {
    /// Scratch space for one formatted line. Reused every instruction, so tracing does
    /// not allocate.
    buffer: [line_buffer_size]u8 = undefined,

    /// Where lines go. `null` formats and discards, which is useful for measuring the
    /// cost of tracing but is otherwise not something you would ask for.
    out: ?*std.Io.Writer = null,

    pub fn emit(
        self: *Tracer,
        step: u64,
        pc: u32,
        raw: u32,
        insn: decode.Instruction,
        regs: *const [16]u32,
    ) void {
        const line = formatLine(&self.buffer, step, pc, raw, insn, regs) catch return;
        if (self.out) |w| w.writeAll(line) catch {};
    }
};

/// Render one trace line into `buf` and return the slice that was written.
///
/// Returns `error.NoSpaceLeft` if the buffer is too small — `std.fmt.bufPrint` never
/// writes past the end of the slice it is given, which is why a fixed buffer is safe here.
pub fn formatLine(
    buf: []u8,
    step: u64,
    pc: u32,
    raw: u32,
    insn: decode.Instruction,
    regs: *const [16]u32,
) error{NoSpaceLeft}![]const u8 {
    // `bufPrint` returns the slice it filled; `written` tracks how much of `buf` is used
    // so the next call can start where the last one stopped.
    var written: usize = 0;

    // `@tagName` turns an enum value into its source-level name at compile time — this is
    // where the `bitwise_and` spelling from the Op enum shows up in the output.
    written += (try std.fmt.bufPrint(buf[written..], "{{\"step\":{d},\"pc\":\"0x{x:0>8}\",\"raw\":\"0x{x:0>6}\",\"len\":{d},\"op\":\"{s}\",\"a\":[", .{
        step,
        pc,
        raw,
        insn.length,
        @tagName(insn.op),
    })).len;

    for (regs, 0..) |value, i| {
        const sep = if (i == 0) "" else ",";
        written += (try std.fmt.bufPrint(buf[written..], "{s}\"0x{x}\"", .{ sep, value })).len;
    }

    written += (try std.fmt.bufPrint(buf[written..], "]}}\n", .{})).len;
    return buf[0..written];
}

// ---------------------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------------------

const testing = std.testing;

test "trace line shape" {
    var buf: [line_buffer_size]u8 = undefined;
    var regs = [_]u32{0} ** 16;
    regs[2] = 0x48;

    const insn = decode.Instruction{ .op = .movi, .length = 3, .t = 2, .imm = 0x48 };
    const line = try formatLine(&buf, 7, 0x4008_0000, 0x4822_A0, insn, &regs);

    try testing.expect(std.mem.startsWith(u8, line, "{\"step\":7,\"pc\":\"0x40080000\","));
    try testing.expect(std.mem.indexOf(u8, line, "\"raw\":\"0x4822a0\"") != null);
    try testing.expect(std.mem.indexOf(u8, line, "\"op\":\"movi\"") != null);
    try testing.expect(std.mem.indexOf(u8, line, "\"a\":[\"0x0\",\"0x0\",\"0x48\",") != null);
    try testing.expect(std.mem.endsWith(u8, line, "]}\n"));
}

test "every register appears exactly once" {
    var buf: [line_buffer_size]u8 = undefined;
    var regs: [16]u32 = undefined;
    for (&regs, 0..) |*reg, i| reg.* = @intCast(i);

    const insn = decode.Instruction{ .op = .nop, .length = 3 };
    const line = try formatLine(&buf, 0, 0, 0x0020F0, insn, &regs);

    var commas: usize = 0;
    for (line) |c| {
        if (c == ',') commas += 1;
    }
    // 5 commas separating the six scalar fields, plus 15 between the 16 registers.
    try testing.expectEqual(@as(usize, 5 + 15), commas);
    try testing.expect(std.mem.indexOf(u8, line, "\"0xf\"]") != null);
}

test "a too-small buffer is an error, not an overrun" {
    var buf: [16]u8 = undefined;
    const regs = [_]u32{0} ** 16;
    const insn = decode.Instruction{ .op = .nop, .length = 3 };
    try testing.expectError(error.NoSpaceLeft, formatLine(&buf, 0, 0, 0, insn, &regs));
}

File: crates/simulatr-core/src/trace.rs
//! Instruction trace: one JSON object per executed instruction.
//!
//! ```text
//! {"step":3,"pc":"0x40080006","raw":"0x0020c0","len":3,"op":"memw","a":["0x0", ...]}
//! ```
//!
//! JSON *Lines* — one self-contained object per line, no enclosing array — so it streams,
//! greps, and pipes into `jq` while the run is still going.
//!
//! Rendered by hand rather than with `serde`: `simulatr-core` has no dependencies by
//! design, and the shape is fixed enough that a serializer would add surface without
//! adding correctness. Formatting into a caller-supplied `String` keeps this pure, so the
//! crate still does no I/O.

use crate::cpu::core::Executed;
use core::fmt::Write as _;

/// Append one trace line, including its terminating newline.
pub fn write_json_line(out: &mut String, step: u64, executed: &Executed, regs: &[u32; 16]) {
    let Executed { pc, word, decoded } = executed;

    // Writing into a String is infallible; the only error `write!` can surface here comes
    // from a Display impl, and none of ours fail.
    let _ = write!(
        out,
        r#"{{"step":{step},"pc":"{pc:#010x}","raw":"0x{word:06x}","len":{len},"op":"{op}","a":["#,
        len = decoded.size.bytes(),
        op = decoded.mnemonic(),
    );

    for (i, value) in regs.iter().enumerate() {
        if i > 0 {
            out.push(',');
        }
        let _ = write!(out, r#""{value:#x}""#);
    }

    out.push_str("]}\n");
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cpu::decode::{Decoded, Inst, InstSize, Reg};

    fn executed(inst: Inst, size: InstSize, pc: u32, word: u32) -> Executed {
        Executed { pc, word, decoded: Decoded { inst, size } }
    }

    #[test]
    fn line_shape() {
        let mut regs = [0u32; 16];
        regs[2] = 0x48;

        let mut out = String::new();
        write_json_line(
            &mut out,
            7,
            &executed(
                Inst::Movi { dst: Reg::new(2), imm: 0x48 },
                InstSize::Wide,
                0x4008_0000,
                0x4822_A0,
            ),
            &regs,
        );

        assert!(out.starts_with(r#"{"step":7,"pc":"0x40080000","#));
        assert!(out.contains(r#""raw":"0x4822a0""#));
        assert!(out.contains(r#""len":3"#));
        assert!(out.contains(r#""op":"movi""#));
        assert!(out.contains(r#""a":["0x0","0x0","0x48","#));
        assert!(out.ends_with("]}\n"));
    }

    #[test]
    fn narrow_instructions_report_their_narrow_mnemonic() {
        let mut out = String::new();
        write_json_line(
            &mut out,
            1,
            &executed(
                Inst::Movi { dst: Reg::new(5), imm: 5 },
                InstSize::Narrow,
                0x4008_000E,
                0x550C,
            ),
            &[0; 16],
        );
        assert!(out.contains(r#""op":"movi.n""#));
        assert!(out.contains(r#""len":2"#));
        assert!(out.contains(r#""raw":"0x00550c""#));
    }

    #[test]
    fn every_register_appears_exactly_once() {
        let mut regs = [0u32; 16];
        for (i, r) in regs.iter_mut().enumerate() {
            *r = i as u32;
        }

        let mut out = String::new();
        write_json_line(
            &mut out,
            0,
            &executed(Inst::Nop, InstSize::Wide, 0, 0x0020F0),
            &regs,
        );

        // Five commas separate the six scalar fields, plus fifteen between the registers.
        assert_eq!(out.matches(',').count(), 5 + 15);
        assert!(out.contains(r#""0xf"]"#));
    }

    #[test]
    fn lines_are_independent_and_appendable() {
        let mut out = String::new();
        let e = executed(Inst::Nop, InstSize::Wide, 0x4008_0000, 0x0020F0);
        write_json_line(&mut out, 1, &e, &[0; 16]);
        write_json_line(&mut out, 2, &e, &[0; 16]);
        assert_eq!(out.lines().count(), 2);
        assert!(out.lines().all(|line| line.starts_with('{') && line.ends_with('}')));
    }
}
